File: src/main/java/org/jukeboxmc/block/behavior/BlockSnowLayer.java
package org.jukeboxmc.block.behavior;

import com.nukkitx.nbt.NbtMap;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.World;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockSnowLayer extends Block {

    public BlockSnowLayer( Identifier identifier ) {
        super( identifier );
    }

    public BlockSnowLayer( Identifier identifier, NbtMap blockStates ) {
        super( identifier, blockStates );
    }

    @Override
    public boolean placeBlock(@NotNull Player player, @NotNull World world, Vector blockPosition, @NotNull Vector placePosition, Vector clickedPosition, Item itemInHand, BlockFace blockFace ) {
        Block block = world.getBlock( blockPosition );

        if (!(world.getBlock( placePosition.subtract( 0, 1, 0 ) ).isSolid())) {
            return false;
        }

        if ( block instanceof BlockSnowLayer blockSnowLayer) {
            if ( blockSnowLayer.getHeight() != 7 ) {
                blockSnowLayer.setHeight( blockSnowLayer.getHeight() + 1 );
                world.setBlock( blockPosition, blockSnowLayer );
            } else {
                block = world.getBlock( placePosition );
                if ( block instanceof BlockSnowLayer ) {
                    blockSnowLayer = (BlockSnowLayer) block;
                    if ( blockSnowLayer.getHeight() != 7 ) {
                        blockSnowLayer.setHeight( blockSnowLayer.getHeight() + 1 );
                        world.setBlock( placePosition, blockSnowLayer );
                    } else {
                        this.setHeight( 0 );
                        world.setBlock( placePosition, this );
                    }
                } else {
                    this.setHeight( 0 );
                    world.setBlock( placePosition, this );
                }
            }
        } else {
            this.setHeight( 0 );
            world.setBlock( placePosition, this );
        }
        return true;
    }

    @Override
    public boolean canBeReplaced( Block block ) {
        if ( block instanceof BlockSnowLayer ) {
            return this.getHeight() != 7;
        } else {
            return false;
        }
    }

    public void setCovered( boolean value ) {
        this.setState( "covered_bit", value ? (byte) 1 : (byte) 0 );
    }

    public boolean isCovered() {
        return this.stateExists( "covered_bit" ) && this.getByteState( "covered_bit" ) == 1;
    }

    public BlockSnowLayer setHeight( int value ) {
        return this.setState( "height", value );
    }

    public int getHeight() {
        return this.stateExists( "height" ) ? this.getIntState( "height" ) : 0;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockSnowLayer.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.BlockType
import org.jukeboxmc.api.block.SnowLayer
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.extensions.toByte
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld

class BlockSnowLayer(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates), SnowLayer {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        var block = world.getBlock(blockPosition)
        val blockDown = world.getBlock(placePosition).getRelative(BlockFace.DOWN)
        if (!blockDown.isSolid() && blockDown.getType() != BlockType.SNOW_LAYER) return false

        if (block is BlockSnowLayer) {
            if (block.getHeight() != 7) {
                block.setHeight(block.getHeight() + 1)
                world.setBlock(blockPosition, block)
            } else {
                block = world.getBlock(placePosition)
                if (block is BlockSnowLayer) {
                    if (block.getHeight() != 7) {
                        block.setHeight(block.getHeight() + 1)
                        world.setBlock(placePosition, block)
                    } else {
                        this.setHeight(0)
                        world.setBlock(placePosition, this)
                    }
                } else {
                    this.setHeight(0)
                    world.setBlock(placePosition, this)
                }
            }
        } else {
            this.setHeight(0)
            world.setBlock(placePosition, this)
        }
        return true
    }

    override fun canBeReplaced(block: JukeboxBlock): Boolean {
        return this.getHeight() < 7
    }

    override fun canPassThrough(): Boolean {
        return true
    }

    override fun isCovered(): Boolean {
        return this.getBooleanState("covered_bit")
    }

    override fun setCovered(value: Boolean): SnowLayer {
        return this.setState("covered_bit", value.toByte())
    }

    override fun getHeight(): Int {
        return this.getIntState("height")
    }

    override fun setHeight(value: Int): SnowLayer {
        return this.setState("height", value)
    }
}